File: VS/lab3/Resource.cpp
#include "Resource.hpp"
#include "WarehouseEntry.hpp"

Resource::Resource(int id, std::string name, MeasureUnit measureUnit, float quantity)
    : WarehouseEntry<float>(id, name), m_measureUnit(measureUnit), m_quantity(quantity) {}

MeasureUnit Resource::GetMeasureUnit() const {
    return m_measureUnit;
}

float Resource::GetQuantity() const {
    return m_quantity;
}

bool Resource::IsEmpty() const {
    return m_quantity <= 0;
}

void Resource::Increase(float quantity) {
    m_quantity += quantity;
}

float Resource::Decrease(float quantity) {
    float taken = std::min(m_quantity, quantity);
    m_quantity -= taken;
    return taken;
}

File: lab8/Java/src/warehouse/resources/Resource.java
package warehouse.resources;

import enums.MaterialType;
import enums.MeasureUnit;
import request.SupplyRequestDetail;
import warehouse.WarehouseEntry;

public abstract class Resource extends WarehouseEntry<Float> {

    Float quantity;
    MeasureUnit measureUnit;

    public static final float MIN_QUANTITY = 0.1f;

    public Resource(int id, String name, MaterialType materialType, MeasureUnit measureUnit, float quantity) {
        super(id, name, materialType);
        if(quantity < MIN_QUANTITY){
            throw new IllegalArgumentException("Недопустимое количество");
        }            
        this.measureUnit = measureUnit;
        this.quantity = quantity;
    }

    public MeasureUnit getMeasureUnit(){
        return measureUnit;
    }

    public float getQuantity(){
        return quantity;
    }

    @Override
    public void increase(Number value){
        quantity+=value.floatValue();
    }

    @Override
    public Float decrease(Number value){
        Float taken = Math.min(quantity, value.floatValue());
        quantity-=taken;
        return taken;
    }
    
    @Override
    public Boolean isEmpty(){
        return quantity < MIN_QUANTITY;
    }

    @Override
    public abstract Boolean isMatches(SupplyRequestDetail detail);
}